traffic_ops/traffic_ops_golang/server/update.go

```diff
@@ -136,29 +141,166 @@ func UpdateHandler(w http.ResponseWriter, r *http.Request) {
+	if !hasConfig_update_time && !hasConfig_apply_time && !hasRevalidate_update_time && !hasRevalidate_apply_time {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("must pass at least one query parameter: 'config_update_time', 'config_apply_time', 'revalidate_update_time', 'revalidate_apply_time'"), nil)
+		return
+	}
+
+	// Validate parameters before attempting to apply them (don't want to partially apply various status before an error)
+	var configUpdateTime time.Time
+	if hasConfig_update_time {
+		configUpdateTime, err = time.Parse(time.RFC3339Nano, config_update_time)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("query parameter 'config_update_time' must be valid RFC3339Nano format"), nil)
+			return
+		}
+	}
+
+	var configApplyTime time.Time
+	if hasConfig_apply_time {
+		configApplyTime, err = time.Parse(time.RFC3339Nano, config_apply_time)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("query parameter 'config_apply_time' must be valid RFC3339Nano format"), nil)
+			return
+		}
+	}
+
+	var revalUpdateTime time.Time
+	if hasRevalidate_update_time {
+		revalUpdateTime, err = time.Parse(time.RFC3339Nano, revalidate_update_time)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("query parameter 'revalidate_update_time' must be valid RFC3339Nano format"), nil)
+			return
+		}
+	}
+
+	var revalApplyTime time.Time
+	if hasRevalidate_apply_time {
+		revalApplyTime, err = time.Parse(time.RFC3339Nano, revalidate_apply_time)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("query parameter 'revalidate_apply_time' must be valid RFC3339Nano format"), nil)
+			return
+		}
+	}
+
+	if hasConfig_update_time {
+		if err = dbhelpers.QueueUpdateForServerWithTime(inf.Tx.Tx, serverID, configUpdateTime); err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("setting config update time: %w", err))
+			return
+		}
+	}
+
+	if hasConfig_apply_time {
+		if err = dbhelpers.SetApplyUpdateForServerWithTime(inf.Tx.Tx, serverID, configApplyTime); err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("setting config apply time: %w", err))
+			return
+		}
+	}
+
+	if hasRevalidate_update_time {
+		if err = dbhelpers.QueueRevalForServerWithTime(inf.Tx.Tx, serverID, revalUpdateTime); err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("setting reval update time: %w", err))
+			return
+		}
+	}
+
+	if hasRevalidate_apply_time {
+		if err = dbhelpers.SetApplyRevalForServerWithTime(inf.Tx.Tx, serverID, revalApplyTime); err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("setting reval apply time: %w", err))
+			return
+		}
+	}
+
+	hostName, _, err := dbhelpers.GetServerNameFromID(inf.Tx.Tx, serverID)
```

Comment:
We don't really _need_ this call. I only use it for the final response. I was trying to be consistent with the response for the previous version of the API. For simplicity we could just return the `idOrName` that was sent as a quick way to reduce the calls to the DB.

Also, I'm given to understand that hostnames are not necessarily unique, even though we treat them as such, which is why I tried to utilize the ServerID in all the new logic instead.

As an aside, I've been theorizing about a data access layer as a way to 1) consolidate all the DB logic into a single spot in the project and 2) as a way to migrate towards utilizing stored procedures. Stored procedures would give us a benefit of reducing back and forth round trips to TO application as well as cached query plans. I believe they could give us a noticeable benefit, especially in high traffic endpoints.

Reply:
> For simplicity we could just return the idOrName that was sent as a quick way to reduce the calls to the DB.

I think that would be fine -- the response can use whatever identifier the client also used.

> data access layer

Indeed, the way we currently access the DB is not ideal -- lots of duplicate (or mostly-duplicate) queries all over the place. This would be quite the undertaking at this point, and I wish we'd been more thoughtful about the data access layer when TO was being rewritten from Perl to Go.

> migrate towards utilizing stored procedures

There are definitely varying opinions on the use of stored procedures. That said, once Cache Config Snapshots are implemented, DB load will be reduced tremendously to the point where we hopefully don't have to worry much about hyper optimization, so we may be able to continue with the simplicity of regular queries.

Reply:
Done